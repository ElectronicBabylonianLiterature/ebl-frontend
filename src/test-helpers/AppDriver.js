import React from 'react'
import { fireEvent, render, waitForElement } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import App from 'App'
import Auth from 'auth/Auth'
import WordRepository from 'dictionary/WordRepository'
import FragmentRepository from 'fragmentarium/FragmentRepository'
import ImageRepository from 'fragmentarium/ImageRepository'
import FragmentService from 'fragmentarium/FragmentService'
import WordService from 'dictionary/WordService'
import TextService from 'corpus/TextService'
import Session from 'auth/Session'
import BibliographyRepository from 'bibliography/BibliographyRepository'
import BibliographyService from 'bibliography/BibliographyService'
import { defaultErrorReporter } from 'ErrorReporterContext'
import createAuth0Config from 'auth/createAuth0Config'

function createApp(api, sessionStore) {
  const auth0Config = createAuth0Config()
  const auth = new Auth(sessionStore, defaultErrorReporter, auth0Config)
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    bibliographyService
  )
  const textService = new TextService(api)

  return (
    <App
      auth={auth}
      wordService={wordService}
      fragmentService={fragmentService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  initialEntries = []
  element = null
  session = null

  constructor(api) {
    this.api = api
  }

  withPath(path) {
    this.initialEntries = [path]
    return this
  }

  withSession() {
    this.session = new Session(
      'accessToken',
      'idToken',
      Number.MAX_SAFE_INTEGER,
      ['write:texts']
    )
    return this
  }

  render() {
    const fakeSessionStore = {
      setSession: session => {
        this.session = session
      },
      clearSession: () => {
        this.session = null
      },
      getSession: () => this.session || new Session('', '', 0, [])
    }

    this.element = render(
      <MemoryRouter initialEntries={this.initialEntries}>
        {createApp(this.api, fakeSessionStore)}
      </MemoryRouter>
    )

    return this
  }

  async waitForText(text) {
    await waitForElement(() => this.element.getByText(text))
  }

  expectTextContent(text) {
    expect(this.element.container).toHaveTextContent(text)
  }

  expectNotInContent(text) {
    expect(this.element.queryByText(text)).toBeNull()
  }

  expectLink(text, expectedHref) {
    expect(this.element.getByText(text)).toHaveAttribute('href', expectedHref)
  }

  expectBreadcrumbs(crumbs) {
    this.expectTextContent(crumbs.join(''))
  }

  expectInputElement(label, expectedValue) {
    expect(this.element.getByLabelText(label).value).toEqual(
      String(expectedValue)
    )
  }

  changeValueByLabel(label, newValue) {
    const input = this.element.getByLabelText(label)
    fireEvent.change(input, { target: { value: newValue } })
  }

  click(text, n = 0) {
    const clickable = this.element.getAllByText(text)[n]
    fireEvent.click(clickable)
    // await wait()
  }
}
