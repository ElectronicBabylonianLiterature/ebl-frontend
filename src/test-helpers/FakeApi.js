import Promise from 'bluebird'

export default class FakeApi {
  gets = []
  client = {
    fetchJson: (path, authenticate) => {
      const expectation = this.gets.find(entry => entry.path === path && entry.authenticate === authenticate)
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchJson: ${path}`))
    },

    fetchBlob: (path, authenticate) => {
      return Promise.reject(new Error(`Unexpected ${authenticate ? 'authenticated' : 'not-authenticated'} fetchBlob: ${path}`))
    }
  }

  expectText (text) {
    this.gets.push({
      path: `/texts/${text.category}/${text.index}`,
      authenticate: true,
      response: text
    })
    return this
  }

  expectStatistics (statistics) {
    this.gets.push({
      path: `/statistics`,
      authenticate: false,
      response: statistics
    })
    return this
  }
}
