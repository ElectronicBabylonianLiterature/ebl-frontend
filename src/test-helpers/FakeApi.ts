import Promise from 'bluebird'
import Word from 'dictionary/domain/Word'

class Expectation {
  method: 'POST' | 'GET' = 'GET'
  path = ''
  authenticate = true
  response = {}
  verify = false
  called = false
  body: object | null = null
  isBlob = false

  constructor(data) {
    Object.assign(this, data)
  }
}

export default class FakeApi {
  private expectations: Expectation[] = []

  readonly client = {
    fetchJson: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchJson: ${path}`
            )
          )
    }),

    postJson: jest.fn().mockImplementation(path => {
      const expectation = this.expectations.find(
        entry => entry.method === 'POST' && entry.path === path && !entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(new Error(`Unexpected postJson: ${path}`))
    }),

    fetchBlob: jest.fn().mockImplementation((path, authenticate) => {
      const expectation = this.expectations.find(
        entry =>
          entry.method === 'GET' &&
          entry.path === path &&
          entry.authenticate === authenticate &&
          entry.isBlob
      )
      return expectation
        ? Promise.resolve(expectation.response)
        : Promise.reject(
            new Error(
              `Unexpected ${
                authenticate ? 'authenticated' : 'not-authenticated'
              } fetchBlob: ${path}`
            )
          )
    })
  }

  expectTexts(texts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts`,
        authenticate: false,
        response: texts,
        verify: true
      })
    )
    return this
  }

  allowText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text
      })
    )
    return this
  }

  expectText(text): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/texts/${text.category}/${text.index}`,
        authenticate: true,
        response: text,
        verify: true
      })
    )
    return this
  }

  expectUpdateManuscripts(text, chapterIndex, manuscripts): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/manuscripts`,
        response: text,
        verify: true,
        body: manuscripts
      })
    )
    return this
  }

  expectUpdateLines(text, chapterIndex, lines): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/texts/${text.category}/${text.index}/chapters/${chapterIndex}/lines`,
        response: text,
        verify: true,
        body: lines
      })
    )
    return this
  }

  expectAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/annotations`,
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectUpdateAnnotations(
    number: string,
    annotationDtos: readonly object[]
  ): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/fragments/${number}/annotations`,
        body: {
          fragmentNumber: number,
          annotations: annotationDtos
        },
        authenticate: true,
        response: { annotations: annotationDtos },
        verify: true
      })
    )
    return this
  }

  expectFragment(fragmentDto): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${fragmentDto._id}`,
        authenticate: true,
        response: fragmentDto,
        verify: true
      })
    )
    return this
  }

  expectPhoto(number: string): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/fragments/${number}/photo`,
        authenticate: true,
        isBlob: true,
        verify: true
      })
    )
    return this
  }

  expectSearchWords(query: string, words: readonly Word[]): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words?query=${encodeURIComponent(query)}`,
        response: words,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/words/${encodeURIComponent(word._id)}`,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  expectUpdateWord(word: Word): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'POST',
        path: `/words/${encodeURIComponent(word._id)}`,
        body: word,
        response: word,
        authenticate: true,
        verify: true
      })
    )
    return this
  }

  allowStatistics(statistics): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/statistics`,
        authenticate: false,
        response: statistics
      })
    )
    return this
  }

  allowImage(file): FakeApi {
    this.expectations.push(
      new Expectation({
        method: 'GET',
        path: `/images/${file}`,
        authenticate: false,
        isBlob: true
      })
    )
    return this
  }

  verifyExpectations(): void {
    const methods = {
      GET: (expectation: Expectation): void =>
        expect(
          expectation.isBlob ? this.client.fetchBlob : this.client.fetchJson
        ).toHaveBeenCalledWith(expectation.path, expectation.authenticate),
      POST: (expectation: Expectation): void =>
        expect(this.client.postJson).toHaveBeenCalledWith(
          expectation.path,
          expectation.body || expect.anything()
        )
    }
    this.expectations
      .filter(expectation => expectation.verify)
      .forEach(expectation => methods[expectation.method](expectation))
  }
}
