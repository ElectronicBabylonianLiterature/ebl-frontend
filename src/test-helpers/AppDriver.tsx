import React from 'react'
import {
  fireEvent,
  render,
  RenderResult,
  act,
  Matcher,
  within
} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import _ from 'lodash'
import App from 'App'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import FragmentRepository from 'fragmentarium/infrastructure/FragmentRepository'
import ApiImageRepository from 'fragmentarium/infrastructure/ImageRepository'
import FragmentService from 'fragmentarium/application/FragmentService'
import WordService from 'dictionary/application/WordService'
import TextService from 'corpus/TextService'
import MemorySession, { Session, guestSession } from 'auth/Session'
import BibliographyRepository from 'bibliography/infrastructure/BibliographyRepository'
import BibliographyService from 'bibliography/application/BibliographyService'
import FragmentSearchService from 'fragmentarium/application/FragmentSearchService'
import { Promise } from 'bluebird'
import { submitForm } from 'test-helpers/utils'
import { eblNameProperty, AuthenticationContext } from 'auth/Auth'

function createApp(api): JSX.Element {
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ApiImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    wordRepository,
    bibliographyService
  )
  const fragmentSearchService = new FragmentSearchService(fragmentRepository)
  const textService = new TextService(api)

  return (
    <App
      wordService={wordService}
      fragmentService={fragmentService}
      fragmentSearchService={fragmentSearchService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  private readonly api
  private initialEntries: string[] = []
  private element: RenderResult | null = null
  private session: Session | null = null

  constructor(api) {
    this.api = api
  }

  getElement(): RenderResult {
    if (this.element) {
      return this.element
    } else {
      throw new Error('getElement called before render.')
    }
  }

  withPath(path): AppDriver {
    this.initialEntries = [path]
    return this
  }

  withSession(): AppDriver {
    this.session = new MemorySession([
      'write:texts',
      'read:fragments',
      'annotate:fragments',
      'read:words'
    ])
    return this
  }

  async render(): Promise<AppDriver> {
    await act(async () => {
      this.element = render(
        <MemoryRouter initialEntries={this.initialEntries}>
          <AuthenticationContext.Provider
            value={{
              login: _.noop,
              logout: _.noop,
              getSession: (): Session => this.session ?? guestSession,
              isAuthenticated: (): boolean => this.session !== null,
              getAccessToken(): Promise<string> {
                throw new Error('Not implemented')
              },
              getUser(): { [eblNameProperty]: string } {
                return { [eblNameProperty]: 'Test' }
              }
            }}
          >
            {createApp(this.api)}
          </AuthenticationContext.Provider>
        </MemoryRouter>
      )
    })

    return this
  }

  async waitForText(text: Matcher): Promise<void> {
    await this.getElement().findByText(text)
  }

  expectTextContent(text: string | RegExp): void {
    expect(this.getElement().container).toHaveTextContent(text)
  }

  expectNotInContent(text: Matcher): void {
    expect(this.getElement().queryByText(text)).toBeNull()
  }

  expectLink(text: Matcher, expectedHref: string): void {
    expect(this.getElement().getByText(text)).toHaveAttribute(
      'href',
      expectedHref
    )
  }

  expectBreadcrumbs(crumbs: readonly string[]): void {
    this.expectTextContent(crumbs.join(''))
  }

  expectBreadcrumb(crumb: string, link: string): void {
    expect(
      within(this.getElement().getByLabelText('breadcrumb')).getByText(crumb)
    ).toHaveAttribute('href', link)
  }

  expectInputElement(label: Matcher, expectedValue: unknown): void {
    expect(
      (this.getElement().getByLabelText(label) as HTMLInputElement).value
    ).toEqual(String(expectedValue))
  }

  changeValueByLabel(label: Matcher, newValue): void {
    const input = this.getElement().getByLabelText(label)
    act(() => {
      fireEvent.change(input, { target: { value: newValue } })
    })
  }

  async submitForm(): Promise<void> {
    await submitForm(this.getElement())
  }

  async click(text: Matcher, n = 0): Promise<void> {
    const clickable = this.getElement().getAllByText(text)[n]
    await act(async () => {
      fireEvent.click(clickable)
    })
  }
}
