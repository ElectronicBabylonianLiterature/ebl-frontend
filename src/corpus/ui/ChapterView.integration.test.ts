import AppDriver from 'test-support/AppDriver'
import FakeApi from 'test-support/FakeApi'
import { produce } from 'immer'

const category = 1
const index = 1
const textDto = {
  category: category,
  index: index,
  name: 'Palm and Vine',
  numberOfVerses: 99,
  approximateVerses: false,
  chapters: [
    {
      classification: 'Ancient',
      stage: 'Neo-Babylonian',
      version: 'A',
      name: 'III',
      order: 3,
      manuscripts: [],
      uncertainFragments: [],
      lines: [],
    },
    {
      classification: 'Ancient',
      stage: 'Old Babylonian',
      version: 'B',
      name: 'The First Chapter',
      order: 1,
      manuscripts: [
        {
          id: 1,
          siglumDisambiguator: '1c',
          museumNumber: 'BM.X',
          accession: 'X.1',
          periodModifier: 'Late',
          period: 'Ur III',
          provenance: 'Nippur',
          type: 'School',
          notes: 'some notes',
          colophon: '1. kur',
          references: [],
        },
      ],
      uncertainFragments: [],
      lines: [],
    },
    {
      classification: 'Ancient',
      stage: 'Old Babylonian',
      version: '',
      name: 'The Second Chapter',
      order: 5,
      manuscripts: [
        {
          id: 1,
          siglumDisambiguator: 'A',
          museumNumber: '',
          accession: '',
          periodModifier: 'Late',
          period: 'Ur III',
          provenance: 'Nippur',
          type: 'School',
          notes: '',
          colophon: '',
          references: [],
        },
        {
          id: 2,
          siglumDisambiguator: 'B',
          museumNumber: '',
          accession: '',
          periodModifier: 'Late',
          period: 'Ur III',
          provenance: 'Nippur',
          type: 'School',
          notes: '',
          colophon: '',
          references: [],
        },
      ],
      uncertainFragments: [],
      lines: [
        {
          number: "1'",
          isBeginningOfSection: false,
          isSecondLineOfParallelism: false,
          variants: [
            {
              reconstruction: 'ideal',
              manuscripts: [
                {
                  manuscriptId: 1,
                  labels: ['iii'],
                  number: 'a+1',
                  atf: 'kur',
                  omittedWords: [],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
const defaultManuscriptDto = {
  id: null,
  siglumDisambiguator: '',
  museumNumber: '',
  accession: '',
  periodModifier: 'None',
  period: 'Neo-Assyrian',
  provenance: 'Nineveh',
  type: 'Library',
  notes: '',
  colophon: '',
  references: [],
}

const defaultLineDto = {
  number: '',
  isBeginningOfSection: false,
  isSecondLineOfParallelism: false,
  variants: [
    {
      reconstruction: '%n ',
      manuscripts: [],
    },
  ],
}

let fakeApi: FakeApi
let appDriver: AppDriver

afterEach(() => {
  fakeApi.verifyExpectations()
})

describe('Diplay chapter', () => {
  const chapter = textDto.chapters[1]
  const chapterTitle = createChapterTitle(chapter)

  beforeEach(async () => {
    await setup(chapter)
  })

  test('Breadcrumbs', () => {
    appDriver.expectBreadcrumbs(['eBL', 'Corpus', chapterTitle])
  })

  test.each([
    ['Classification', 'classification'],
    ['Stage', 'stage'],
    ['Version', 'version'],
    ['Name', 'name'],
  ])('%s', (label, property) => {
    appDriver.expectInputElement(label, chapter[property])
  })

  describe('Manuscript', () => {
    const manuscript = chapter.manuscripts[0]

    test.each([
      ['Siglum', 'siglumDisambiguator', 'b'],
      ['Museum Number', 'museumNumber', 'BM.X2'],
      ['Accession', 'accession', 'X.2'],
      ['Period modifier', 'periodModifier', 'Early'],
      ['Period', 'period', 'Hellenistic'],
      ['Provenance', 'provenance', 'Borsippa'],
      ['Type', 'type', 'Commentary'],
      ['Notes', 'notes', 'more notes'],
    ])('%s', async (label, property, newValue) => {
      fakeApi.expectUpdateManuscripts(
        textDto,
        1,
        produce(textDto.chapters[1].manuscripts, (draft) => ({
          manuscripts: [
            {
              ...draft[0],
              [property]: newValue,
            },
          ],
          uncertainFragments: textDto.chapters[1].uncertainFragments,
        }))
      )
      const value = manuscript[property]
      const expectedValue = value.name ? value.name : value
      appDriver.expectInputElement(label, expectedValue)
      await appDriver.changeValueByLabel(label, newValue)
      appDriver.expectInputElement(label, newValue)
      await appDriver.click('Save manuscripts')
    })
  })
})

describe('Add manuscript', () => {
  const chapter = textDto.chapters[0]

  beforeEach(async () => {
    await setup(chapter)
  })

  test.each([
    ['Siglum', 'siglumDisambiguator', ''],
    ['Museum Number', 'museumNumber', ''],
    ['Accession', 'accession', ''],
    ['Period modifier', 'periodModifier', 'None'],
    ['Period', 'period', 'Neo-Assyrian'],
    ['Provenance', 'provenance', 'Nineveh'],
    ['Type', 'type', 'Library'],
    ['Notes', 'notes', ''],
  ])('%s', async (label, property, expectedValue) => {
    const manuscript = {
      ...defaultManuscriptDto,
      [property]: expectedValue,
      id: 1,
    }
    fakeApi.expectUpdateManuscripts(textDto, 0, {
      manuscripts: [manuscript],
      uncertainFragments: [],
    })
    await appDriver.click('Add manuscript')
    appDriver.expectInputElement(label, expectedValue)
    await appDriver.click('Save manuscripts')
  })
})

test('Uncertain Fragments', async () => {
  const chapter = textDto.chapters[0]
  const museumNumber = 'X.1'
  const label = 'Museum Number'

  await setup(chapter)
  await appDriver.click('Add fragment')
  await appDriver.changeValueByLabel(label, museumNumber)
  appDriver.expectInputElement(label, museumNumber)
  await appDriver.click('Save manuscripts')

  fakeApi.expectUpdateManuscripts(textDto, 0, {
    manuscripts: chapter.manuscripts,
    uncertainFragments: [museumNumber],
  })
})

describe('Lines', () => {
  const chapter = textDto.chapters[2]
  const line = chapter.lines[0]

  beforeEach(async () => {
    await setup(chapter, false)
    await appDriver.click('Lines')
  })

  test.each([['Number', 'number', '2']])(
    '%s',
    async (label, property, newValue) => {
      fakeApi.expectUpdateLines(textDto, 2, {
        lines: [
          produce(textDto.chapters[2].lines[0], (draft) => {
            draft[property] = newValue
          }),
        ],
      })
      const expectedValue = line[property]
      appDriver.expectInputElement(label, expectedValue)
      await appDriver.changeValueByLabel(label, newValue)
      appDriver.expectInputElement(label, newValue)
      await appDriver.click('Save lines')
    }
  )

  test.each([
    ['second line of parallelism', 'isSecondLineOfParallelism'],
    ['beginning of a section', 'isBeginningOfSection'],
  ])('%s', async (label, property) => {
    fakeApi.expectUpdateLines(textDto, 2, {
      lines: [
        produce(textDto.chapters[2].lines[0], (draft) => {
          draft[property] = !draft[property]
        }),
      ],
    })
    const expectedValue = line[property]
    expectedValue
      ? appDriver.expectChecked(label)
      : appDriver.expectNotChecked(label)
    await appDriver.click(label)
    expectedValue
      ? appDriver.expectNotChecked(label)
      : appDriver.expectChecked(label)
    await appDriver.click('Save lines')
  })
})

describe('Add line', () => {
  const chapter = textDto.chapters[1]

  beforeEach(async () => {
    await setup(chapter)
    await appDriver.click('Lines')
  })

  test.each([['Number', 'number']])('%s', async (label, property) => {
    fakeApi.expectUpdateLines(textDto, 1, { lines: [defaultLineDto] })
    await appDriver.click('Add line')
    appDriver.expectInputElement(label, defaultLineDto[property])
    await appDriver.click('Save lines')
  })
})

test('Import chapter', async () => {
  const chapter = textDto.chapters[1]
  fakeApi.expectImportChapter(textDto, 1, '1. kur')
  await setup(chapter)
  await appDriver.click('Import')
  await appDriver.click('Save')
})

describe('Chapter not found', () => {
  const chapter = textDto.chapters[1]
  const chapterName = 'Unknown Chapter'

  beforeEach(async () => {
    fakeApi = new FakeApi().allowText(textDto)
    appDriver = await new AppDriver(fakeApi.client)
      .withSession()
      .withPath(createChapterPath(chapter.stage, chapterName))
      .render()

    await appDriver.waitForText(`Edit ${textDto.name}`)
  })

  test('Error message', () => {
    appDriver.expectTextContent(`Chapter not found.`)
  })
})

async function setup(chapter, expectText = true) {
  fakeApi = expectText
    ? new FakeApi().expectText(textDto)
    : new FakeApi().allowText(textDto)
  appDriver = await new AppDriver(fakeApi.client)
    .withSession()
    .withPath(createChapterPath(chapter.stage, chapter.name))
    .render()
  await appDriver.waitForText(`Edit ${createChapterTitle(chapter)}`)
}

function createChapterPath(stage: string, name: string) {
  return `/corpus/${encodeURIComponent(category)}/${encodeURIComponent(
    index
  )}/${encodeURIComponent(stage)}/${encodeURIComponent(name)}`
}

function createChapterTitle(chapter) {
  return `${textDto.name} ${chapter.stage} ${chapter.name}`
}
