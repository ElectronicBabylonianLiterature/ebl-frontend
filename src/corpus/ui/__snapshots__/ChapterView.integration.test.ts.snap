// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diplay chapter Snapshot 1`] = `
<div>
  <header
    class="Header"
  >
    <nav
      class="navbar navbar-expand-md navbar-light"
    >
      <div
        class="container"
      >
        <span
          class="navbar-brand"
        >
          <a
            href="https://www.lmu.de"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              class="Header__corporate-logo img-fluid"
              src="LMU_Logo.svg"
            />
          </a>
        </span>
        <a
          class="active navbar-brand"
          href="/"
          title="electronic Babylonian Literature (eBL)"
        >
          <h1
            class="Header__title"
          >
            <span
              class="Header__title-main"
            >
              electronic
              <br />
              Babylonian
              <br />
              Literature
            </span>
            <small
              class="Header__title-abbreviation"
            >
              eBL
            </small>
          </h1>
        </a>
        <div
          class="navbar-collapse collapse"
          id="Header-2"
        >
          <div
            class="mx-auto navbar-nav"
          >
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/signs"
                href="/signs"
              >
                Signs
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/dictionary"
                href="/dictionary"
              >
                Dictionary
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="active nav-link"
                data-rb-event-key="/corpus"
                href="/corpus"
              >
                Corpus
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/fragmentarium"
                href="/fragmentarium"
              >
                Fragmentarium
              </a>
            </div>
            <div
              class="nav-item"
            >
              <a
                class="nav-link"
                data-rb-event-key="/bibliography"
                href="/bibliography"
              >
                Bibliography
              </a>
            </div>
          </div>
          <span
            class="navbar-text"
          >
            <button
              class="btn btn-outline-secondary btn-sm"
              type="button"
            >
              Logout 
              Test
            </button>
          </span>
        </div>
        <button
          aria-controls="Header-2"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
      </div>
    </nav>
  </header>
  <section
    class="App-content"
  >
    <header
      class="App-content__header"
    >
      <nav
        aria-label="breadcrumb"
      >
        <ol
          class="breadcrumb"
        >
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/"
            >
              eBL
            </a>
          </li>
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/corpus"
            >
              Corpus
            </a>
          </li>
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/corpus/L"
            >
              L
            </a>
          </li>
          <li
            class="breadcrumb-item active"
          >
            <a
              href="/corpus/L/1/1"
            >
              I.1 Lo sikfagdeh ir lino oveagfes bohhi watil te zulreziv weemwin fodiric ipagiiki wugob coh bowko iz owaofoece zuh.
            </a>
          </li>
          <li
            aria-current="page"
            class="breadcrumb-item active"
          >
            Chapter Old Babylonian The Only Chapter
          </li>
        </ol>
      </nav>
      <div
        class="float-right btn-toolbar"
        role="toolbar"
      >
        <div
          class="btn-group"
          role="group"
        >
          <div
            class="dropdown btn-group"
            role="group"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-outline-primary"
              type="button"
            >
              Go to
            </button>
          </div>
          <a
            class="btn btn-outline-primary"
            href="/corpus/L/1/1/Old%20Babylonian/The%20Only%20Chapter/edit"
          >
            <i
              class="fas fa-edit"
            />
             Edit
          </a>
        </div>
      </div>
      <h2>
        Lo sikfagdeh ir lino oveagfes bohhi watil te zulreziv weemwin fodiric ipagiiki wugob coh bowko iz owaofoece zuh.
        <br />
        <small>
          Chapter
           
          The Only Chapter
           
          <span>
            <span>
              Pofo jarikat vowhukod voawka pigiz jato milatas sik desit ronvojjiw ga muvcaeni ke.
            </span>
          </span>
        </small>
      </h2>
    </header>
    <table
      class="chapter-display"
    >
      <tbody>
        <tr>
          <td
            class="chapter-display__intertext"
            colspan="4"
          >
            (
            <span>
              <span>
                Na pur gi tufse lero hel patedofo hiusiwal otlucvi fac vokifel umihineni mi gepzor puji famih.
              </span>
            </span>
            )
          </td>
        </tr>
        <tr
          class="chapter-display__line chapter-display__line--is-second-line-of-parallelism chapter-display__line--is-beginning-of-section"
        >
          <td
            class="chapter-display__line-number"
          >
            1
          </td>
          <td
            colspan="1"
          >
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__AkkadianWord Transliteration__AkkadianWord--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__ValueToken"
                >
                  <span
                    class=""
                  >
                    kur-kur
                  </span>
                </span>
                <sup />
              </span>
            </span>
          </td>
          <td
            class="chapter-display__line-number"
          >
            1
          </td>
          <td
            class="chapter-display__translation"
          >
            <div>
              <span>
                Weg bocuvaor fite vanupu zopab sun rofu wapzelaf uc uhoev ficdogpih wujfoj komtedzi ver.
              </span>
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="chapter-display__intertext"
            colspan="4"
          >
            (
            <span>
              <span>
                Zenkitu kuklot lohid etoek geezu ve diluwgar ki zopvonfe beb ki alpemhi cucvugkez cir ujbi ude.
              </span>
            </span>
            )
          </td>
        </tr>
        <tr
          class="chapter-display__line"
        >
          <td
            class="chapter-display__line-number"
          >
            2
          </td>
          <td
            colspan="1"
          >
            <span
              class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
            >
               
            </span>
            <span
              class="Transliteration__AkkadianWord Transliteration__AkkadianWord--AKKADIAN"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__ValueToken"
                >
                  <span
                    class=""
                  >
                    kur-kur
                  </span>
                </span>
                <sup />
              </span>
            </span>
          </td>
          <td
            class="chapter-display__line-number"
          >
            2
          </td>
          <td
            class="chapter-display__translation"
          >
            <div>
              <span>
                Wijada pacopavi lepoaze ahawa ezis nofa ihelvok imaesu zu paduis zocuzmev oc.
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;
